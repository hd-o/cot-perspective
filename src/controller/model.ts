import { config } from '@/model/config'

export class ModelController {
  constants = config
}
