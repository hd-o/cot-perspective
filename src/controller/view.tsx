import { Controller } from '@/controller'
import { COTData, FormattedCSVData, MarketsData, TraderCategory } from '@/model/types'
import { Template } from '@/view/template'

interface RenderExchangeProps {
  data: COTData
  exchanges: string[]
}

interface RenderMarketProps {
  data: COTData
  exchange: string
  exchanges: string[]
  markets: string[]
  marketsData: MarketsData
}

interface RenderTemplateProps {
  data: COTData
  exchanges: string[]
  marketData: FormattedCSVData[]
  markets: string[]
  selections: {
    exchange: string
    market: string
    traderCategory: TraderCategory
  }
}

export class ViewController {
  private get _constants () {
    return this.ctrl.model.constants
  }

  constructor (private readonly ctrl: Controller) {}

  renderExchange (props: RenderExchangeProps) {
    return (exchange: string) => {
      const marketsData = props.data[exchange]
      const markets = this.ctrl.data.getSortedKeys(marketsData)
      markets.forEach(this.renderMarket({
        ...props,
        exchange,
        markets,
        marketsData,
      }))
    }
  }

  renderMarket (props: RenderMarketProps) {
    return (market: string) => {
      const marketData = props.marketsData[market]
      for (const traderCategory of this._constants.traderCategories) {
        this.renderTemplate({
          ...props,
          marketData,
          selections: {
            exchange: props.exchange,
            market,
            traderCategory,
          },
        })
      }
    }
  }

  async renderPages () {
    const data = await this.ctrl.data.getData({
      year: new Date().getFullYear(),
      minimumEntries: this._constants.averagePeriod,
    })
    const exchanges = this.ctrl.data.getSortedKeys(data)
    console.log(`• Rendering pages for ${exchanges.length} exchanges`)
    exchanges.forEach(this.renderExchange({ data, exchanges }))
  }

  renderTemplate (props: RenderTemplateProps) {
    console.log('• Processing template data for ', props.selections)
    const template = (
      <Template
        dropDownsData={{
          data: props.data,
          markets: props.markets,
          exchanges: props.exchanges,
          traderCategories: this._constants.traderCategories,
          ...props.selections,
        }}
        tableData={{
          averagePeriod: this._constants.averagePeriod,
          values: props.marketData.map(
            this.ctrl.data.processTableData(props.selections.traderCategory),
          ),
        }}
      />
    )
    this.ctrl.pkg.node.fs.writeFileSync(
      `${this._constants.buildPath}/${this.ctrl.file.getPageId(props.selections)}.html`,
      `<!doctype html> \n${this.ctrl.pkg.reactDom.renderToString(template)}`,
    )
  }
}
