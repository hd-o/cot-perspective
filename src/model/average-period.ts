/** Quantity of periods to calculate the average of COT values */
export const averagePeriod = 12
