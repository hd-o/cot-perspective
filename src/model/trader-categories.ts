/** COT categories */
export const traderCategories = ['Commercial', 'Noncommercial'] as const
