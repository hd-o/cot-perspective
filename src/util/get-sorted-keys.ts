type GetSortedKeys = (d: Record<string, unknown>) => string[]

/** Returns sorted keys array from given object */
export const getSortedKeys: GetSortedKeys = (data) => Object.keys(data).sort()
