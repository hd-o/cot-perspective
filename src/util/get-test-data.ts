import { testData } from '../model/test-data'
import { COTData } from '../model/types'
import { GetDataConfig } from './get-data'
import { Use } from './resolve-container'

type GetTestData = (data?: GetDataConfig) => Promise<COTData>

export const useGetTestData: Use<GetTestData> = () => {
  return async () => testData
}
