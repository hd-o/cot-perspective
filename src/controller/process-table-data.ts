import { TraderCategory } from '../model/trader-categories'
import { CSVData, FormattedCSVData } from '../model/types'
import { Use } from './resolve-container'

type ProcessTableData =
  (c: TraderCategory) =>
  (v: FormattedCSVData) => Array<string|number>

/**
 * For a given trader category, return a function that prepares
 * an array of data to be used in table/view rendering
 */
export const useProcessTableData: Use<ProcessTableData> = () => {
  return (selectedTraderCategory) => (values) => {
    const longs = Number(
      values[`${selectedTraderCategory} Positions-Long (All)` as keyof CSVData]
    )
    const shorts = Number(
      values[`${selectedTraderCategory} Positions-Short (All)` as keyof CSVData]
    )
    const netPositions = longs - shorts
    return [
      values['As of Date in Form YYYY-MM-DD'],
      longs,
      shorts,
      Number(
        values[`Change in ${selectedTraderCategory}-Long (All)` as keyof CSVData]
      ),
      Number(
        values[`Change in ${selectedTraderCategory}-Short (All)` as keyof CSVData]
      ),
      values[`% of OI-${selectedTraderCategory}-Long (All)` as keyof CSVData],
      values[`% of OI-${selectedTraderCategory}-Short (All)` as keyof CSVData],
      netPositions,
    ]
  }
}
