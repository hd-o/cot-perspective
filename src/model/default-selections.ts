import { DropDownSelections } from '../view/page-dropdowns'
import { traderCategories } from './trader-categories'

/** Index pages, or test default page selection */
export const defaultSelections: DropDownSelections = {
  exchange: 'CHICAGO MERCANTILE EXCHANGE',
  market: 'EURO FX',
  traderCategory: traderCategories[1]
}
