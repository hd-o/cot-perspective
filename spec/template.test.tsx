import { shallow } from 'enzyme'
import React from 'react'
import { averagePeriod } from '../src/model/average-period'
import { defaultSelections } from '../src/model/default-selections'
import { traderCategories } from '../src/model/trader-categories'
import { getSortedKeys } from '../src/util/get-sorted-keys'
import { getTestData } from '../src/util/get-test-data'
import { processTableData } from '../src/util/process-table-data'
import { Template } from '../src/view/template'

/** Verify correct Template rendering of test data */
test('Template snapshot', async () => {
  const testData = await getTestData()
  const marketsData = testData[defaultSelections.exchange]
  const marketData = marketsData[defaultSelections.market]
  const template = (
    <Template
      dropDownsData={{
        data: testData,
        exchanges: getSortedKeys(testData),
        markets: getSortedKeys(testData[defaultSelections.exchange]),
        traderCategories,
        ...defaultSelections
      }}
      tableData={{
        averagePeriod,
        values: marketData.map(
          processTableData(defaultSelections.traderCategory)
        )
      }}
    />
  )
  expect(shallow(template)).toMatchSnapshot()
})
