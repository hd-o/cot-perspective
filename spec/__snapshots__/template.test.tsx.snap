// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template snapshot 1`] = `ShallowWrapper {}`;
